### DISCUSSION TITLE
Starting with Supabase Migrations in Existing Environments

### DISCUSSION THREAD

**Discussion by ianbaker615** [supabase/supabase#29545] [Questions]
Hello Supabase Community,

I am trying to start using Supabase migrations to manage and track changes to databases across different environments (dev, staging, and prod). Up until now, all changes to the database tables have been made manually via the Supabase UI. I just created dev and staging as copies of prod about a week ago.

Objective:
- Start using migrations to make database changes from this point forward.
- Ensure smooth transition without errors when applying migrations.

Current Issues:
1. Baseline Migration:
- Created a baseline migration capturing the current state.
- Encountering errors such as ERROR: type "http_header" already exists (SQLSTATE 42710), indicating pre-existing objects are causing conflicts.
- AI keeps suggesting using DROP IF EXISTS in the migration file, but any baseline migration that I generate is thousands of lines long so it doesn't seem practical.

2. Applying New Migrations:
- Created small test migrations for subsequent changes.
- These migrations are also failing, often because they conflict with the existing state of objects in the database.

Questions:
1. How can I effectively create a baseline migration for my existing schema that won't fail due to pre-existing objects?
2. What is the best approach to ensure my environment databases (dev, staging, prod) are in sync and handle existing objects gracefully?
3. Any recommended practices or steps when transitioning from manual database changes to using migrations in Supabase?

I've attempted to follow typical migration practices and ensuring a consistent starting point, but I'm still running into issues. Any help, guidance, or shared experiences would be greatly appreciated.
Thank you!

**Comment by nicelion** [comment]
Hello fellow Ian.

Currently having the same issues.

Were you able to find a solution?

**Comment by empz** [comment]
Same issue here, want to start using declarative schemas and these NOTICES popup every time. They doesn't make any sense to me as those are tables managed by Supabase, not by our own migrations.

**Comment by alexrabin** [comment]
I'm also having this issue. This has been open for over a year. Can someone from supabase take a look at this please? @SupaDavid @supa-alex @dventimisupabase

**Reply by erondpowell** [reply]
This is such a friction point, and there's so many loose pieces not covered by the docs. 

I've spent several days in analysis paralysis. If I can get any stakeholder to hop on a 45-60min google meet with me and talk over the recommended dev-to-prod approaches, I will write the guides and shoot tutorial videos for free.

CC @alexrabin if you got something worked out, and wanna share, I'd LOVE to polish it up & share it public.
Also CC **@ANYBODY- READING** — _and @SupaDavid @supa-alex @dventimisupabase because you are tagged above._

**Reply by alexrabin** [reply]
I still don't have a fully working solution. It involves using GitHub actions and database merges and and deleting the merge history each time. It's kind of a b*tch to be honest. I would love to see a better solution.

**Reply by erondpowell** [reply]
Rats! Okay..

**Comment by farmhutsoftwareteam** [comment]
supabase db dump -f
  supabase/migrations/00000000000001_baseline.sql

**Reply by erondpowell** [reply]
@farmhutsoftwareteam yep, great for the database. But there's peripheral stuff.
- storage config
- auth config
- how does branching work? It doesn't configure auth and you have to do it manually of via mgmt api? What are its other limits? 

There's so many micro things that don't seem to be covered by the database migration.

If you have an full coverage migrations workflow, and you (allmost) never need to configure stuff in the production project, I'd love to hear it.

**Comment by brunocek** [comment]
I am on the same boat: https://github.com/orgs/supabase/discussions/41806
